### PR TITLE
Add deployment for zip requests

### PR DESCRIPTION
**Pull request by RMeharg**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -183,5 +184,6 @@ func (p *Page) Page(description string) {
   em.AddHandler(p, "deploy.start")
   em.AddHandler(p, "deploy.success")
   em.AddHandler(p, "deploy.failure")
+	em.AddHandler(p, "deploy.error")
```
Comment (trayo) [inline README.md]: The tab here isn't needed.

File: artifetcher/artifetcher.go [hunk 2]
```diff
@@ -60,6 +61,7 @@ func (a *Artifetcher) Fetch(url, manifest string) (string, error) {
 	}
 	defer response.Body.Close()
 
+	println(response.Body)
```
Comment (trayo) [inline artifetcher/artifetcher.go]: Was this left on accident? If not we should use the logger.

File: README.md [hunk 3]
```diff
@@ -123,6 +123,7 @@ With Deployadactyl you can optionally register event handlers to perform any add
 |`deploy.start`|[DeployEventData](structs/deploy_event_data.go)|Before deployment starts
 |`deploy.success`|[DeployEventData](structs/deploy_event_data.go)|When a deployment succeeds
 |`deploy.failure`|[DeployEventData](structs/deploy_event_data.go)|When a deployment fails
+|`deploy.error`|[DeployEventData](structs/deploy_event_data.go)|When a deployment fails
```
Comment (trayo) [inline README.md]: We should probably use a different description for `deploy.error`

File: artifetcher/artifetcher.go [hunk 4]
```diff
@@ -84,3 +86,37 @@ func (a *Artifetcher) Fetch(url, manifest string) (string, error) {
...
+	defer a.FileSystem.Remove(zipFile.Name())
+
+	f := bytes.NewReader(requestBody)
+	if _, err = io.Copy(zipFile, f); err != nil {
+		return "", errors.Errorf("%s: %s", cannotWriteResponseToFile, err)
+	}
+
+	a.Log.Debug("Fetching local file: %s", zipFile.Name())
+
+	//Create temp dir
```
Comment (trayo) [inline artifetcher/artifetcher.go]: Is the comment needed? The following doesn't look weird enough to comment on.

File: controller/controller_test.go [hunk 5]
```diff
@@ -91,262 +78,157 @@ var _ = Describe("Controller", func() {
...
-		Context("username and password are provided", func() {
-			It("accepts the request with a 200 OK", func() {
-				eventManager.EmitCall.Returns.Error = nil
-				deployer.DeployCall.Write.Output = "push succeeded"
-				deployer.DeployCall.Returns.Error = nil
-
-				By("setting authenticate to true")
-				controller.Config.Environments[environment] = config.Environment{Authenticate: true}
-
+	Context("the controller receives a request that has a mime type application/json", func() {
```
Comment (trayo) [inline controller/controller_test.go]: These `Context` and `Describe` should be flipped.  

> You use Describe blocks to describe the individual behaviors of your code and Context blocks to excercise those behaviors under different circumstances. 

From the [Ginkgo docs](https://onsi.github.io/ginkgo/#organizing-specs-with-containers-describe-and-context) docs

File: controller/controller_test.go [hunk 6]
```diff
@@ -91,262 +78,157 @@ var _ = Describe("Controller", func() {
...
+			})
 		})
 	})
 
-	Describe("deployment output", func() {
-		It("shows the user deployment info properties", func() {
-			eventManager.EmitCall.Returns.Error = nil
-			deployer.DeployCall.Returns.Error = nil
-
+	Context("the controller receives a request that is not a recognised mime type", func() {
```
Comment (trayo) [inline controller/controller_test.go]: We use recogni _**z**_ ed. ;)

File: controller/deployer/deployer.go [hunk 7]
```diff
@@ -2,47 +2,143 @@
...
-func (d Deployer) Deploy(deploymentInfo S.DeploymentInfo, out io.Writer) (err error) {
-	var appPath string
+func (d Deployer) Deploy(req *http.Request, environmentName, org, space, appName string, out io.Writer) (err error, statusCode int) {
+	var (
+		appPath                string
+		deploymentInfo         = S.DeploymentInfo{}
+		environments           = d.Config.Environments
+		authenticationRequired = environments[environmentName].Authenticate
+		deployEventData        = S.DeployEventData{}
+	)
```
Comment (trayo) [inline controller/deployer/deployer.go]: Could use a new line here after the `var` block

File: controller/deployer/deployer.go [hunk 8]
```diff
@@ -2,47 +2,143 @@
...
+	deploymentInfo, err = getDeploymentInfo(req.Body)
+	if err != nil {
+		println(err.Error())
```
Comment (trayo) [inline controller/deployer/deployer.go]: We should use the `logger` or the `out` buffer

File: controller/deployer/deployer.go [hunk 9]
```diff
@@ -2,47 +2,143 @@
...
+	deploymentInfo.Manifest = string(m)
+
+	defer func() (error, int) {
+		deployFinishEvent := S.Event{
+			Type: "deploy.finish",
+			Data: deployEventData,
+		}
+
+		eventErr := d.EventManager.Emit(deployFinishEvent)
+		if eventErr != nil {
```
Comment (trayo) [inline controller/deployer/deployer.go]: Do we not want to print `eventErr`?

**Comment by RMeharg** [comment]
Updated for review @trayo

File: controller/controller_test.go [hunk 10]
```diff
@@ -91,262 +78,151 @@ var _ = Describe("Controller", func() {
...
-		Context("username and password are provided", func() {
-			It("accepts the request with a 200 OK", func() {
-				eventManager.EmitCall.Returns.Error = nil
-				deployer.DeployCall.Write.Output = "push succeeded"
-				deployer.DeployCall.Returns.Error = nil
-
-				By("setting authenticate to true")
-				controller.Config.Environments[environment] = config.Environment{Authenticate: true}
-
+	Describe("the controller receives a request that has a mime type application/json", func() {
```
Comment (trayo) [inline controller/controller_test.go]: The text "the controller receives a request that has" isn't needed in this `Describe`.

File: controller/controller_test.go [hunk 11]
```diff
@@ -91,262 +78,151 @@ var _ = Describe("Controller", func() {
...
-		Context("username and password are provided", func() {
-			It("accepts the request with a 200 OK", func() {
-				eventManager.EmitCall.Returns.Error = nil
-				deployer.DeployCall.Write.Output = "push succeeded"
-				deployer.DeployCall.Returns.Error = nil
-
-				By("setting authenticate to false")
-				controller.Config.Environments[environment] = config.Environment{Authenticate: false}
-
+	Describe("the controller receives a request that has a mime type application/zip", func() {
```
Comment (trayo) [inline controller/controller_test.go]: The text "the controller receives a request that has" isn't needed in this `Describe`.

File: controller/controller_test.go [hunk 12]
```diff
@@ -91,262 +78,151 @@ var _ = Describe("Controller", func() {
...
-				eventManager.EmitCall.Returns.Error = nil
-				deployer.DeployCall.Write.Output = "push succeeded"
-				deployer.DeployCall.Returns.Error = nil
-
-				By("setting authenticate to true")
-				controller.Config.Environments[environment] = config.Environment{Authenticate: true}
-
+	Describe("the controller receives a request that has a mime type application/json", func() {
+		Context("successful deployments without missing properties with a remote artifact url", func() {
+			It("deploys successfully with no error message and a status code of 200 OK", func() {
```
Comment (trayo) [inline controller/controller_test.go]: "deploys successfully" is redundant given the context says "successful deployments".

File: controller/controller_test.go [hunk 13]
```diff
@@ -91,262 +78,151 @@ var _ = Describe("Controller", func() {
...
-			It("accepts the request with a 200 OK", func() {
-				eventManager.EmitCall.Returns.Error = nil
-				deployer.DeployCall.Write.Output = "push succeeded"
-				deployer.DeployCall.Returns.Error = nil
-
-				By("setting authenticate to true")
-				controller.Config.Environments[environment] = config.Environment{Authenticate: true}
-
+	Describe("the controller receives a request that has a mime type application/json", func() {
+		Context("successful deployments without missing properties with a remote artifact url", func() {
```
Comment (trayo) [inline controller/controller_test.go]: I feel like this should read as "successful deployments without missing properties **_and_** with a remote artifact url"

File: controller/controller_test.go [hunk 14]
```diff
@@ -91,262 +78,151 @@ var _ = Describe("Controller", func() {
...
+				Expect(resp.Body).To(ContainSubstring("deploy successful"))
 			})
 		})
 
-		Context("when username and password are not provided", func() {
-			It("rejects the request with a 401 unauthorized", func() {
-				By("setting authenticate to true")
-				controller.Config.Environments[environment] = config.Environment{Authenticate: true}
-
+		Context("failed deployments", func() {
```
Comment (trayo) [inline controller/controller_test.go]: The `Context` here should be "when it fails to build the application". Then next `It` should be something like `It("returns and error")`.

File: controller/controller_test.go [hunk 15]
```diff
@@ -91,262 +78,151 @@ var _ = Describe("Controller", func() {
...
+			It("fails to build the application", func() {
```
Comment (trayo) [inline controller/controller_test.go]: This `It` should read like "It("returns and error")."

File: controller/controller_test.go [hunk 16]
```diff
@@ -91,262 +78,151 @@ var _ = Describe("Controller", func() {
...
-						"artifact_url": "%s",
-						"manifest": "%s"
-					}`,
-					artifactURL,
-					base64Manifest,
-				))
-
-				req, err := http.NewRequest("POST", apiURL, jsonBuffer)
+		Context("failed deployments", func() {
+			It("request is empty", func() {
```
Comment (trayo) [inline controller/controller_test.go]: This `It` probably needs a little more info

File: controller/deployer/deployer_test.go [hunk 17]
```diff
@@ -199,30 +215,251 @@ var _ = Describe("Deployer", func() {
...
+
+				err, statusCode := deployer.Deploy(req, environmentName, org, space, appName, wd, zipRequest, buffer)
+				Expect(err).To(MatchError("an error occurred in the deploy.start event"))
+				Expect(statusCode).To(Equal(500))
+				Expect(buffer).To(ContainSubstring("event error"))
+			})
+		})
+	})
+
+	Describe("Common Functionality", func() {
```
Comment (trayo) [inline controller/deployer/deployer_test.go]: I think this `Describe`.

**Pull request merged**